src/app/ui/dock.cpp

```diff
@@ -398,22 +440,98 @@ bool Dock::onProcessMessage(ui::Message* msg)
 
     case kMouseMoveMessage: {
       if (hasCapture()) {
+        const gfx::Point& pos = static_cast<MouseMessage*>(msg)->position();
+
+        if (m_dropzonePlaceholder)
+          m_dropzonePlaceholder->setGhostPosition(pos);
+
         if (m_hit.sideIndex >= 0) {
-          const gfx::Point pos = static_cast<MouseMessage*>(msg)->position();
+          if (!display()->bounds().contains(pos) ||
+              (m_hit.widget && m_hit.widget->parent() &&
+               !m_hit.widget->parent()->bounds().contains(pos)))
+            break; // Do not handle anything outside bounds.
+
           gfx::Size& sz = m_sizes[m_hit.sideIndex];
+          gfx::Size minSize(16 * guiscale(), 16 * guiscale());
+
+          if (m_hit.widget) {
+            minSize.w = std::max(m_hit.widget->minSize().w, minSize.w);
+            minSize.h = std::max(m_hit.widget->minSize().h, minSize.h);
+          }
 
           switch (m_hit.sideIndex) {
-            case kTopIndex:    sz.h = (m_startSize.h + pos.y - m_startPos.y); break;
-            case kBottomIndex: sz.h = (m_startSize.h - pos.y + m_startPos.y); break;
-            case kLeftIndex:   sz.w = (m_startSize.w + pos.x - m_startPos.x); break;
-            case kRightIndex:  sz.w = (m_startSize.w - pos.x + m_startPos.x); break;
+            case kTopIndex: sz.h = std::max(m_startSize.h + pos.y - m_startPos.y, minSize.h); break;
+            case kBottomIndex:
+              sz.h = std::max(m_startSize.h - pos.y + m_startPos.y, minSize.h);
+              break;
+            case kLeftIndex:
+              sz.w = std::max(m_startSize.w + pos.x - m_startPos.x, minSize.w);
+              break;
+            case kRightIndex:
+              sz.w = std::max(m_startSize.w - pos.x + m_startPos.x, minSize.w);
+              break;
           }
 
           layout();
           Resize();
         }
-        else if (m_hit.dockable) {
+        else if (m_hit.dockable && m_dragging) {
           invalidate();
+
+          auto* parentDock = dynamic_cast<Dock*>(m_hit.widget->parent());
+          ASSERT(parentDock);
+
+          if (!parentDock)
+            break;
+
+          if (!m_dropzonePlaceholder)
+            m_dropzonePlaceholder.reset(new DropzonePlaceholder(m_hit.widget, pos));
+
+          auto dockedAt = parentDock->whichSideChildIsDocked(m_hit.widget);
+          const auto& bounds = parentDock->bounds();
+
+          if (!bounds.contains(pos))
+            break; // Do not handle anything outside the bounds of the dock.
+
+          const int BUFFER_ZONE =
+            std::max(12 * guiscale(), std::min(m_hit.widget->size().w, m_hit.widget->size().h));
```

Comment:
Probably I'd change this to `kBufferZone`:
```suggestion
          const int kBufferZone =
            std::max(12 * guiscale(), std::min(m_hit.widget->size().w, m_hit.widget->size().h));
```
As all capital letters are reserved more for macros or enums.